### PR TITLE
Doubles && replacing buildFunc func type getter && short circuiting `ABIInfo::computeInfo`

### PR DESCRIPTION
**Pull request by lanza**
This is three commits.

* The first adds support for doubles for call. (This is what we agreed to get rid of, the third commit does this.)
* The second replaces `buildFunc`'s function type getter with the machinery modeled from CodeGen.
* The third opts out of all the ABI consideration by just short cutting `ABIInfo::computeInfo`.

### CODE REVIEW

File: clang/lib/CIR/TargetInfo.cpp [hunk 1]
```diff
@@ -112,7 +112,37 @@ ABIArgInfo X86_64ABIInfo::getIndirectResult(QualType Ty,
   assert(false && "NYI");
 }
 
+static bool testIfIsVoidTy(QualType Ty) {
```
Comment (bcardosolopes) [inline clang/lib/CIR/TargetInfo.cpp:115]: How about `isVoidTy(...)`? It seems that we've implemented similar logic somewhere else in the builder, can you double check this isn't already available?

File: clang/lib/CIR/TargetInfo.cpp [hunk 2]
```diff
@@ -112,7 +112,37 @@ ABIArgInfo X86_64ABIInfo::getIndirectResult(QualType Ty,
...
+  const auto *BT = Ty->getAs<BuiltinType>();
+  if (!BT)
+    return false;
+
+  BuiltinType::Kind k = BT->getKind();
+  return k == BuiltinType::Void;
+}
+
 void X86_64ABIInfo::computeInfo(CIRGenFunctionInfo &FI) const {
+  // Top leevl CIR has unlimited arguments and return types. Lowering for ABI
```
Comment (bcardosolopes) [inline clang/lib/CIR/TargetInfo.cpp:125]: *level

File: clang/lib/CIR/TargetInfo.cpp [hunk 3]
```diff
@@ -112,7 +112,37 @@ ABIArgInfo X86_64ABIInfo::getIndirectResult(QualType Ty,
...
+  }
+  auto RetTy = FI.getReturnType();
+  if (testIfIsVoidTy(RetTy))
+    FI.getReturnInfo() = ABIArgInfo::getIgnore();
+  else
+    FI.getReturnInfo() = ABIArgInfo::getDirect(CGT.ConvertType(RetTy));
+
+  return;
+
+  // TODO:
```
Comment (bcardosolopes) [inline clang/lib/CIR/TargetInfo.cpp:143]: With the assert we probably don't need this `TODO`

File: clang/lib/CIR/TargetInfo.cpp [hunk 4]
```diff
@@ -112,7 +112,37 @@ ABIArgInfo X86_64ABIInfo::getIndirectResult(QualType Ty,
...
+static bool testIfIsVoidTy(QualType Ty) {
+  const auto *BT = Ty->getAs<BuiltinType>();
+  if (!BT)
+    return false;
+
+  BuiltinType::Kind k = BT->getKind();
+  return k == BuiltinType::Void;
+}
+
 void X86_64ABIInfo::computeInfo(CIRGenFunctionInfo &FI) const {
```
Comment (bcardosolopes) [inline clang/lib/CIR/TargetInfo.cpp:124]: If we are assuming everything is direct right now, why does the code path needs to get here? Is there a way to shortcut this at an upper level? If not, this should probably be calling another function that's shared to other potential targets as well?

**Review by bcardosolopes — APPROVED** [review]
My OCD hits hard with a capital case "GetFunctionType", but I understand that's baggage from LLVM codegen. We already changed so much, I wonder if this is worth changing at some point. In the meantime, LGTM.

**Review by bcardosolopes — APPROVED** [review]
This looks fine for now, so LGTM. I think we need to step back a bit and think more about what ClangIR should look like w.r.t. to these ABI infos, and think less about what LLVM is doing / mimic. One important question to ask is: what mimicking LLVM brings in the table for the current type of analysis we care about?

What if an argument is tagged with a `#direct`, or `#extend` attribute instead? These attributes can be described in terms of any ABI, and similar for CXXABI. Example of how a future version of CIR could look like:

```
// Extend here is defined according to the ABI to be used
#extend = { cir.float<32> : "sse_use80fp" } 

func @c(%arg0: cir.int<u,32> {#direct}, %arg1: cir.float<32> {#extend}) -> f64 {
...
```

**Review by bcardosolopes — APPROVED** [review]
I hope we get to rename `GetFunctionType` to lower case at some point :)

LGTM

**Comment by lanza** [comment]
> I hope we get to rename `GetFunctionType` to lower case at some point :)

I think we should just add a `.clang-tidy` file to `CIR` and abide by it. I tried but utterly failed at keeping up with the dual naming setup and I'm doing both incorrectly :P. What are your thoughts on this?

**Comment by bcardosolopes** [comment]
> I think we should just add a `.clang-tidy` file to `lib/CIR` and `include/clang/CIR` and abide by it. I tried but utterly failed at keeping up with the dual naming setup and I'm doing both incorrectly :P. What are your thoughts on this?

Seems like a good idea, I'm failing all the time too!

**Pull request closed without merging**